### ISSUE TITLE
Supervised edge embedding calculation using SAGEConv (GraphSAGE) 

### ISSUE THREAD

**Issue by rmtr** [rusty1s/pytorch_geometric#1377]
## ❓ Questions & Help

Thanks a lot for this awesome library!

I am trying to use SAGEConv to extract node embeddings from unseen graphs (inductive) that was trained in a supervised manner using my own loss function. Training will be done using thousands of small graphs (anode count of around 750 per graph with average node degree around 5) with high dimensional node features (about 256 features per node). The embeddings will then be used in a binary classification network using standard binary cross entropy loss. I am a complete beginner to all these. So, please bear with me.
My questions are as follows:

1. Following is the rough sketch I'm going to follow to build the model. I would like to know whether there are any issues with the approach. For now, Dataset will be a single graph (one batch is one graph) and later, will try to extend it to multiple graphs.

<pre>class MyGraphSAGE(torch.nn.Module):
    def __init__(self,FeatureDim, Dataset):
        super(MyGraphSAGE,self).__init__()
        self.Dataset = Datset
        self.conv1 = SAGEConv(FeatureDim, 32)
        self.conv2 = SAGEConv(32,32) #i.e. I expect to get 32 dimension embedding

    def forward(self,x,EdgeIndex):
        x, EdgeIndex = x,EdgeIndex
        x = self.conv1(x, EdgeIndex)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        x = self.conv2(x, EdgeIndex)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        return MyLossFunction(x)

    def MyLossFunction(x)
        # Use self.Dataset and x to calculate loss and return it
</pre>

Optimisation  parameters

<pre>
model = MyGraphSAGE(256, Dataset)
model = model.to(device)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
optimiser = torch.optim.Adam(model.parameters(),lr=0.001, weight_decay=0.01)
</pre>

2. I am not exactly sure how I can write a training process for the above model. With my limited understanding, it might be something like this.

<pre>
def train():
    model.train()
    optimizer.zero_grad()
    # As far as I understand, I have to call the custom loss function here and do the gradient descend. 
    # But have no idea how to do so for a custom loss function 
    optimiser.step()
</pre>

Any help is greatly appreciated.

**Comment by rusty1s** [comment]
Hi,

1. This looks totally fine. You can use the `normalize` flag in `SAGEConv` for normalization, too. However, in general, I find that L2-normalization does not work so well in `SAGEConv`, and it should definitely be removed before calling the loss function. In addition, loss computation should be moved outside the model's forward pass.

2. You need to call your model with the input, compute the loss, and backpropagate gradients:
```python
def train(data):
    model.train()
    optimizer.zero_grad()
    out = model(data.x, data.edge_index)
    loss = MyLossFunction(x, data.y)
    loss.backward()
    optimizer.step()
```

**Comment by rmtr** [comment]
Thank you very much for the quick response. I will have a go at this and report back.
I may have some questions later for which I need some guidance.

**Comment by rmtr** [comment]
@rusty1s 
I have a question regarding creating custom datasets that do not fit in to memory and loading them properly (and some other questions as well). Currently, I'm constructing the data loader using `Data` and `DataLoader` in `torch_geometric.data`. The code looks like below.

<pre>
from torch_geometric.data import Data, DataLoader

# Model
class MyGraphSAGE(torch.nn.Module):
    def __init__(self,FeatureDim,):
        super(MyGraphSAGE,self).__init__()
        self.conv1 = SAGEConv(FeatureDim, 32)
        self.conv2 = SAGEConv(32,32) #i.e. I expect to get 32 dimension embedding

    def forward(self,x,EdgeIndex):
        x, EdgeIndex = x,EdgeIndex
        x = self.conv1(x, EdgeIndex)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        x = self.conv2(x, EdgeIndex)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        return x

# Dataset creation
def create_Data_Objects(self):
    # Some preprocessing code
    # GraphData is a numpy ndarray holding node features (x) and node labels (y)
    # EdgeList is a 2 * num_nodes numpy ndarray
    for Number_of_Graphs_in_the_Dataset:
        x = torch.tensor(GraphData[:, 0:-1], dtype=torch.float)
        y = torch.tensor(GraphData[:,-1])
        edge_index = torch.tensor(EdgeList)

        GData = Data(x=x, y=y, edge_index=edge_index)
        Train_Data_list.append(GData)
    return Train_Data_list

# Loss function
def LossFunction(Loss, Embedding, GT, Adjacency):
    # Return the calculated loss

# Training function
def train(numEpoch, data):
    model.train()
    Loss = 1E-10

    for epoch in numEpoch:
        for GraphData in data:
            GraphData.to(device)
            optimiser.zero_grad()
            out = model(GraphData.x, GraphData.edge_index)
            Loss = LossFunction(Loss, out, GraphData.y, GraphData.Adjacency)
            Loss.backward()
            optimiser.step()

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

Train_Data_list = create_Data_Objects()
TrainLoader = DataLoader(Train_Data_list, batch_size=1)

model = MyGraphSAGE(258)
model = model.to(device)
optimiser = torch.optim.Adam(model.parameters(),lr=0.001, weight_decay=0.01)
train(TrainLoader.dataset)
</pre>

My question is that whether my strategy of creating PyG dataset and DataLoader correct and is TrainLoader properly utilised?
Could you also please take some time to comment on the whether this code looks okay in general as well.
Thanks in advance!
